### DISCUSSION TITLE
Mono requirement or pseudo-formula

### DISCUSSION THREAD

**Discussion by alexreg** [Homebrew/discussions#3555] [Writing Formulae/Casks]
The formula `mono` and the casks `mono-mdk` and `mono-mdk-for-visual-studio` are of course mutually exclusive and provide essentially the same software, albeit at times slightly different versions. It would seem highly desirable to create some sort of requirement or pseudo-formula that is satisfied by any of the three being installed (so long as the version is compatible). As it is right now, some formulas cannot be installed because a given formula depends on a specific one of these three formulas/casks, though it really doesn't make a difference.

Thoughts on what to do about the above would be appreciated. If we can decide on a solution, I would be happy to create a PR.

**Comment by carlocab** [comment]
We have the same situation for the various JDK formulae/casks.

There was a time when Java dependencies were a `Requirement` that could be fulfilled by any Java installation, even if it was not installed via Homebrew.

This resulted in regular bug reports about things not working, because it meant that users were using configurations that were not tested in CI.

Adding flexibility for the Mono dependency is likely going to lead to a similar situation, so I don’t think that’s something we want to do, at least for formulae in Homebrew/core. (You’ll also notice that no dependencies are fulfilled by casks there, and that’s intentional.) You can do whatever you like in an external tap.

**Comment by alexreg** [comment]
@carlocab Thanks for the response. It largely makes sense. Do you however have a recommendation of what to do in the case a formula depends on `mono` but I have `mono-mdk` installed? I suppose I could use `--ignore-dependencies`, but then subsequent updates would fail.

**Reply by carlocab** [reply]
You could `brew unlink mono`. That should stick until it's next reinstalled or upgraded. (Arguably it should stick after that too, but `brew` doesn't know how to handle that.)

**Reply by alexreg** [reply]
Hmm, that seems less than ideal. Do you think we could make that option stick? Or have `--ignore-dependencies` stick? Or some new option?

**Reply by carlocab** [reply]
`--ignore-dependencies` is an unsupported developer flag, so we'd probably sooner remove it than expand it by making it stick.

Making `brew unlink` stick has been on my wish list for a while, but I haven't had the time to look into it. I'd be happy to review a PR that implements it.

**Reply by alexreg** [reply]
Sounds good. I'll have a look at that.

Also, I do think `--ignore-dependencies` is useful for testing formulas during creation, regardless.

**Reply by carlocab** [reply]
Yes, I don't think `--ignore-dependencies` is going away. I was just emphasising how unlikely it was that we'd support making it persist across reinstalls/upgrades.

**Comment by MikeMcQuaid** [comment]
IMO a better fix for this would be to avoid manual `brew unlink`ing but instead have a DSL to define formulae and casks to have overlapping links and somehow decide which one gets linked.

**Reply by alexreg** [reply]
@MikeMcQuaid What are the advantages over the manual approach, that you see? The advantage of the manual approach is that the user can choose which one he links, and all subsequent attempts to install conflicting formulas will succeed in installation but fail in linkage.

**Reply by MikeMcQuaid** [reply]
The manual approach is not at all visible to users and requires each individual user to see errors installing one or both of these before they take action. 

Also, your fix has a non-trivial chance of breaking things for users who unlink at some point and wonder why later other formulae that have assumed links then break.

**Reply by alexreg** [reply]
I agree that visibility to users is a slight issue... However, I think the behaviour of formula remaining in their previous linkage state across reinstalls & upgrades is both totally intuitive and natural, in any case. The question is more whether we want a layer on top of that for managing "overlapping" formulas, for which you make the case.

**Reply by MikeMcQuaid** [reply]
> However, I think the behaviour of formula remaining in their previous linkage state across reinstalls & upgrades is both totally intuitive and natural, in any case.

I disagree. It's neither if/when it breaks things.

> I agree that visibility to users is a slight issue

This is a big deal. If we're going to fix this issue, let's do it properly.

**Reply by alexreg** [reply]
Is it not bad practice for formulas to assume their dependencies are linked?

**Reply by SMillerDev** [reply]
If it was your solution wouldn't work either. Since that depends on a cask taking the link location of a formula

**Reply by alexreg** [reply]
@SMillerDev Whom are you addressing?

**Reply by MikeMcQuaid** [reply]
> Is it not bad practice for formulas to assume their dependencies are linked?

Quite possibly. We don't audit or test against it, though, so it's impossible to tell how many are assuming this.

**Reply by alexreg** [reply]
@MikeMcQuaid I see. Well, your point is certainly stronger in that case. I personally will be relatively happy as long as there is some sort of solution to this issue from a user perspective. Would your approach allow the user to choose which of a set of overlapping formulas gets linked, or would it be by some sort of priority list?
